File: frontend/src/types/patient.ts
import { BaseEntity } from "./common";
import { Title } from "./title";

export interface Patient extends BaseEntity {
  title_id?: number;
  name: string;
  ln?: string;
  gender?: "Male" | "Female" | string;
  cid?: string;
  hn?: string;
  birth_date?: string; // ISO Date (YYYY-MM-DD)

  // 🚩 ฟิลด์พิเศษจาก @property ใน Backend
  age_display?: string;

  // Relationships
  title?: Title;

  // Timestamps (สืบทอดมาจาก BaseEntity ถ้าคุณตั้งค่าไว้)
  created_at?: string;
  updated_at?: string;
}

/**
 * สำหรับใช้ในหน้าค้นหา หรือ Table ที่ต้องการข้อมูลสรุป
 */
export interface PatientListItem extends Patient {
  surgical_cases_count?: number;
}

File: backend/app/models/patient.py
from sqlalchemy import Column, Integer, String, Date, DateTime, ForeignKey
from dateutil.relativedelta import relativedelta
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.database import Base


class Patient(Base):
    __tablename__ = "patients"

    id = Column(Integer, primary_key=True, index=True)
    title_id = Column(Integer, ForeignKey("titles.id"), nullable=True)
    name = Column(String, index=True, nullable=False)
    ln = Column(String, index=True, nullable=True)
    gender = Column(String, nullable=True)
    cid = Column(String, unique=True, index=True, nullable=True)
    birth_date = Column(Date, nullable=True)

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    # Relationships
    title = relationship("Title")

    surgical_cases = relationship(
        "SurgicalCase",
        back_populates="patient",
        cascade="all, delete-orphan",  # (Optional) ลบคนไข้แล้วลบเคสทั้งหมด
    )

    gyne_cytology_cases = relationship("GyneCytologyCase", back_populates="patient")
    nongyne_cytology_cases = relationship("NongyneCytologyCase", back_populates="patient")

    @property
    def age_display(self):
        if not self.birth_date:
            return "-"

        from datetime import date
        # คำนวณเทียบกับวันปัจจุบัน หรือวันรับเคส
        diff = relativedelta(date.today(), self.birth_date)

        if diff.years >= 2:
            return f"{diff.years} Y"
        elif diff.years >= 1:
            return f"{diff.years} Y {diff.months} M"
        elif diff.months >= 1:
            return f"{diff.months} M {diff.days} D"
        else:
            return f"{diff.days} D"
